File: src/runtime-tools.ts
import { Script, createContext } from "node:vm";
import type { JsonValue } from "./protocol.js";
import type { Context, Plugin } from "./runtime.js";

interface DynamicPluginDefinition {
  pluginId: string;
  name: string;
  purpose: string;
  code: string;
  dispose?: () => void;
}

export function runtimeToolsPlugin(): Plugin {
  return {
    name: "runtime-tools",
    setup(context) {
      const definitions = new Map<string, DynamicPluginDefinition>();
      let nextPluginId = 1;

      context.effect(() => {
        for (const definition of [...definitions.values()].reverse()) definition.dispose?.();
        definitions.clear();
      });

      context.contributePrompt(
        "runtime-evolution-guide",
        "Inspect the current Context, define a small Cordis Plugin when the task needs a new capability, run it, use its tools, then stop or undefine it after the experiment.",
      );
      context.registerTool({
        name: "cordis_inspect",
        description: "Inspect mounted plugins, services, tools, Prompt contributions, listeners, and relations.",
        inputSchema: { type: "object", properties: {}, additionalProperties: false },
        execute: () => context.inspect() as unknown as JsonValue,
      });
      context.registerTool({
        name: "cordis_define",
        description: "Define a new Cordis Plugin from a JavaScript function body that returns the Plugin.",
        inputSchema: {
          type: "object",
          properties: {
            name: { type: "string", minLength: 1 },
            purpose: { type: "string", minLength: 1 },
            code: { type: "string", minLength: 1 },
          },
          required: ["name", "purpose", "code"],
          additionalProperties: false,
        },
        execute(input) {
          const { name, purpose, code } = readDefinition(input);
          compilePlugin(code);
          const pluginId = `dyn-${nextPluginId}`;
          nextPluginId += 1;
          definitions.set(pluginId, { pluginId, name, purpose, code });
          return { ok: true, pluginId, name, status: "defined" };
        },
      });
      context.registerTool({
        name: "cordis_run",
        description: "Run one previously defined Cordis Plugin in the current Context.",
        inputSchema: pluginIdSchema(),
        async execute(input) {
          const pluginId = readPluginId(input);
          const definition = definitions.get(pluginId);
          if (!definition) return { ok: false, error: `Unknown dynamic Plugin: ${pluginId}` };
          if (definition.dispose) return { ok: false, error: `${pluginId} is already running.` };
          const plugin = instantiatePlugin(definition);
          definition.dispose = await context.mount(plugin);
          return {
            ok: true,
            pluginId,
            status: "running",
            tools: context.inspect().tools.map((tool) => tool.name),
          };
        },
      });
      context.registerTool({
        name: "cordis_stop",
        description: "Stop a running dynamic Plugin while retaining its definition.",
        inputSchema: pluginIdSchema(),
        execute(input) {
          const pluginId = readPluginId(input);
          const definition = definitions.get(pluginId);
          if (!definition) return { ok: false, error: `Unknown dynamic Plugin: ${pluginId}` };
          if (!definition.dispose) return { ok: false, error: `${pluginId} is not running.` };
          definition.dispose();
          delete definition.dispose;
          return { ok: true, pluginId, status: "stopped" };
        },
      });
      context.registerTool({
        name: "cordis_undefine",
        description: "Stop a dynamic Plugin if needed and remove its definition.",
        inputSchema: pluginIdSchema(),
        execute(input) {
          const pluginId = readPluginId(input);
          const definition = definitions.get(pluginId);
          if (!definition) return { ok: false, error: `Unknown dynamic Plugin: ${pluginId}` };
          definition.dispose?.();
          definitions.delete(pluginId);
          return { ok: true, pluginId, status: "undefined" };
        },
      });
    },
  };
}

function compilePlugin(code: string): Script {
  return new Script(`(() => {\n"use strict";\n${code}\n})()`);
}

function instantiatePlugin(definition: DynamicPluginDefinition): Plugin {
  const value = compilePlugin(definition.code).runInContext(createContext(Object.create(null)), {
    timeout: 100,
  }) as unknown;
  if (
    typeof value !== "object" ||
    value === null ||
    typeof (value as { apply?: unknown }).apply !== "function"
  ) {
    throw new Error("Dynamic code must return a Cordis Plugin with apply(context).");
  }
  const dynamic = value as { apply(context: Context): void | Promise<void> };
  return {
    name: `dynamic:${definition.name}`,
    setup: (context) => dynamic.apply(context),
  };
}

function pluginIdSchema(): Record<string, unknown> {
  return {
    type: "object",
    properties: { pluginId: { type: "string", pattern: "^dyn-[1-9][0-9]*$" } },
    required: ["pluginId"],
    additionalProperties: false,
  };
}

function readDefinition(input: JsonValue): { name: string; purpose: string; code: string } {
  return input as { name: string; purpose: string; code: string };
}

function readPluginId(input: JsonValue): string {
  return (input as { pluginId: string }).pluginId;
}

File: python_harness/runtime_tools.py
"""Chapter 5: define, run, stop, and undefine dynamic plugins."""

from __future__ import annotations

from types import SimpleNamespace
from typing import Any, Awaitable, Callable

from .runtime import Context, Plugin

Tool = SimpleNamespace


def make_tool(
    parameters: dict[str, Any],
    execute: Callable[[dict[str, Any]], Awaitable[Any]],
) -> Tool:
    """Bundle a JSON-Schema parameters object with an async execute function.

    The agent loop validates arguments against parameters, then awaits
    execute(arguments), so every registered tool needs this shape.
    """
    return Tool(parameters=parameters, execute=execute)


def runtime_tools_plugin() -> Plugin:
    """A plugin exposing cordis_* tools for dynamic plugin experiments."""

    def setup(context: Context) -> None:
        definitions: dict[str, dict[str, Any]] = {}
        next_plugin_id = 1

        def cleanup() -> None:
            for definition in reversed(list(definitions.values())):
                dispose = definition.get("dispose")
                if dispose is not None:
                    dispose()
            definitions.clear()

        async def inspect_runtime(arguments: dict[str, Any]) -> dict[str, Any]:
            return context.inspect()

        async def define(arguments: dict[str, Any]) -> dict[str, Any]:
            name = arguments["name"]
            code = arguments["code"]
            compile(code, f"<dynamic:{name}>", "exec")
            nonlocal next_plugin_id
            plugin_id = f"dyn-{next_plugin_id}"
            next_plugin_id += 1
            definitions[plugin_id] = {
                "plugin_id": plugin_id,
                "name": name,
                "purpose": arguments["purpose"],
                "code": code,
            }
            return {"ok": True, "pluginId": plugin_id, "name": name, "status": "defined"}

        async def run(arguments: dict[str, Any]) -> dict[str, Any]:
            plugin_id = arguments["pluginId"]
            definition = definitions.get(plugin_id)
            if definition is None:
                return {"ok": False, "error": f"unknown dynamic plugin: {plugin_id}"}
            if definition.get("dispose") is not None:
                return {"ok": False, "error": f"{plugin_id} is already running"}
            plugin = instantiate_plugin(definition)
            definition["dispose"] = context.mount(plugin)
            return {
                "ok": True,
                "pluginId": plugin_id,
                "status": "running",
                "tools": list(context.tools),
            }

        async def stop(arguments: dict[str, Any]) -> dict[str, Any]:
            plugin_id = arguments["pluginId"]
            definition = definitions.get(plugin_id)
            if definition is None:
                return {"ok": False, "error": f"unknown dynamic plugin: {plugin_id}"}
            dispose = definition.get("dispose")
            if dispose is None:
                return {"ok": False, "error": f"{plugin_id} is not running"}
            dispose()
            definition["dispose"] = None
            return {"ok": True, "pluginId": plugin_id, "status": "stopped"}

        async def undefine(arguments: dict[str, Any]) -> dict[str, Any]:
            plugin_id = arguments["pluginId"]
            definition = definitions.get(plugin_id)
            if definition is None:
                return {"ok": False, "error": f"unknown dynamic plugin: {plugin_id}"}
            dispose = definition.get("dispose")
            if dispose is not None:
                dispose()
            definitions.pop(plugin_id, None)
            return {"ok": True, "pluginId": plugin_id, "status": "undefined"}

        context.effect(cleanup)
        context.contribute_prompt(
            "Inspect the current Context, define a small Cordis Plugin when the task needs a new capability, run it, use its tools, then stop or undefine it after the experiment.",
        )
        context.register_tool("cordis_inspect", make_tool({"type": "object", "properties": {}, "additionalProperties": False}, inspect_runtime))
        context.register_tool("cordis_define", make_tool(
            {
                "type": "object",
                "properties": {
                    "name": {"type": "string", "minLength": 1},
                    "purpose": {"type": "string", "minLength": 1},
                    "code": {"type": "string", "minLength": 1},
                },
                "required": ["name", "purpose", "code"],
                "additionalProperties": False,
            },
            define,
        ))
        context.register_tool("cordis_run", make_tool(plugin_id_schema(), run))
        context.register_tool("cordis_stop", make_tool(plugin_id_schema(), stop))
        context.register_tool("cordis_undefine", make_tool(plugin_id_schema(), undefine))

    return SimpleNamespace(name="runtime-tools", setup=setup)


def instantiate_plugin(definition: dict[str, Any]) -> Plugin:
    """Execute the stored code and require a plugin_factory() returning a Plugin."""
    namespace: dict[str, Any] = {}
    exec(compile(definition["code"], f"<dynamic:{definition['name']}>", "exec"), namespace)
    factory = namespace.get("plugin_factory")
    if not callable(factory):
        raise RuntimeError("dynamic code must define plugin_factory() returning a Plugin")
    plugin = factory()
    if not hasattr(plugin, "name") or not callable(getattr(plugin, "setup", None)):
        raise RuntimeError("plugin_factory() must return a Plugin with name and setup(context)")
    return plugin


def plugin_id_schema() -> dict[str, Any]:
    return {
        "type": "object",
        "properties": {"pluginId": {"type": "string", "pattern": r"^dyn-[1-9][0-9]*$"}},
        "required": ["pluginId"],
        "additionalProperties": False,
    }
